db/cron.c

```diff
@@ -0,0 +1,614 @@
+/*
+   Copyright 2019 Bloomberg Finance L.P.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+#include "cron.h"
+#include "cron_systable.h"
+
+/**
+ * Primitive cron job that monitors a ordered list of epoch marked events,
+ * each having a callback function associated
+ * Sleep until next event is to be triggered
+ *
+ */
+
+struct cron_sched {
+    pthread_t tid; /* pthread id of the thread owning this cron structure */
+    pthread_cond_t cond; /* locking and signaling */
+    pthread_mutex_t mtx;
+    int running; /* marked under mtx lock, signaling a event processing if set
+                  */
+    LISTC_T(struct cron_event) events; /* list of events */
+    LINKC_T(struct cron_sched) lnk;    /* link the cron schedulers */
+    sched_if_t impl;
+};
+
+typedef struct cron_scheds {
+    LISTC_T(struct cron_sched) scheds; /* all global schedulers */
+    pthread_rwlock_t rwlock;
+} cron_scheds_t;
+
+static cron_scheds_t crons;
+pthread_mutex_t _crons_mtx = PTHREAD_MUTEX_INITIALIZER;
+
+static void *_cron_runner(void *arg);
+static int _queue_event(cron_sched_t *sched, int epoch, FCRON func, void *arg1,
+                        void *arg2, void *arg3, uuid_t *source_id,
+                        struct errstat *err);
+
+void init_cron(void)
+{
+    listc_init(&crons.scheds, offsetof(struct cron_sched, lnk));
+    Pthread_rwlock_init(&crons.rwlock, NULL);
+}
+
+/**
+ * Add a new event to a scheduler.
+ * Create a scheduler if none exists yet
+ *
+ */
+cron_sched_t *cron_add_event(cron_sched_t *sched, const char *name, int epoch,
```

Comment:
Want to split this into 2 functions?  One to create a scheduler, one to add an event to an existing scheduler?